coreax/solvers/recombination.py

```diff
@@ -0,0 +1,570 @@
+    algorithm for all but one scenario, the computation of the null space basis. Because
+    the zero-weighted nodes still exist in the node matrix, they influence the SVD and
+    yield an erroneous null space basis.
+
+    We ameliorate this problem by setting the zero-weighted nodes equal (co-linear) to
+    the largest weighted node (an arbitrary but consistent choice). Because the nodes
+    are now co-linear to each other and the largest weighted node, we know that at least
+    all but one of them can be safely eliminated by the recombination procedure. Thus,
+    the nodes become effectively "invisible" to the elimination procedure.
+
+    The only caveat is that we don't know which of the equal nodes will be retained post
+    elimination. To handle this, we keep an index (reference) from the zero-weighted
+    nodes to the largest weighted node, and we redistribute the largest weight equally
+    over all the "co-linearized" nodes (preserving the CoM and allowing any node to be
+    eliminated).
+
+    :param nodes: The nodes to co-linearize
+    :param weights: The weights to apply the co-linearization correction to
+    :return: The co-linearized nodes, corrected weights, and co-linearized-to-original
+        reference indices.
+    """
+    max_index = jnp.argmax(weights)
+    non_zero_weights_mask = weights > 0
+    zero_weights_mask = 1 - non_zero_weights_mask
+    n_zeros = zero_weights_mask.sum()
+    weights = weights.at[max_index].divide(n_zeros + 1)
+    indices = jnp.arange(weights.shape[0])
+    indices *= non_zero_weights_mask
+    indices += zero_weights_mask * max_index
+    return nodes[indices], weights[indices], indices
+
+
+# pylint: disable=line-too-long
+# Credit: https://github.com/patrick-kidger/lineax/blob/9b923c8df6556551fedc7adeea7979b5c7b3ffb0/lineax/_solver/svd.py#L67  # noqa: E501
+# for the rank determination code.
+# pylint: enable=line-too-long
+def _resolve_null_basis(
+    nodes: Shaped[Array, "n m"],
+    rcond: Union[float, None] = None,
+) -> tuple[Shaped[Array, "n n"], int]:
+    r"""
+    Resolve the largest left null space basis, and its rank, for passed the node matrix.
+
+    By largest left null space basis, we mean the null space basis under the assumption
+    that the rank of the null space is maximal (assumed to be ``n``). If the rank is not
+    maximal, then only the first :math:`n - m^\prime` basis vectors will be actual basis
+    vectors for the null space (where :math:`m^\prime` is the rank of the node matrix).
+    The remaining "basis" vectors can, and should, be ignored in upstream computations
+    by using the left null space rank value as a cut-off index.
+
+    :param nodes: Matrix of nodes (m-vectors) whose null space is to be determined
+    :param rcond: The relative condition number of the Matrix of nodes
+    :return: The largest left null space basis and its rank, for the passed node matrix.
+    """
+    q, s, _ = jsp.linalg.svd(nodes, full_matrices=True)
```

Comment:
Just a questions, could randomised SVD be used here to achieve significant speedup?

Reply:
I belive so, or at least I can't think of an immediately obvious reason why not. I'm not sure how it would relate to the randomized recombination method described here though https://www.cs.ox.ac.uk/people/alessandro.abate/publications/cCOA20.pdf.